File: src/com/easy/LongestHarmoniousSubsequence_594.java
/**
 * 
 */
package com.easy;

import java.util.HashMap;
import java.util.Iterator;

/**
 * @author Chris Lin
 * @version 1.0
 */
public class LongestHarmoniousSubsequence_594 {
    // 2019年7月19日
    // Use Recursive
    // Time Complexity: O(n)
    // Space Complexity:O(n)
    // Runtime: 32 ms, faster than 46.02%
    // Memory Usage: 40 MB, less than 97.06%
    public int findLHS(int[] nums) {
        int mResult = 0;
        HashMap<Integer, Integer> mMap = new HashMap<>();
        for (int num : nums) {
            mMap.put(num, mMap.getOrDefault(num, 0) + 1);
        }
        Iterator<Integer> mIter = mMap.keySet().iterator();
        while (mIter.hasNext()) {
            int mTmpNum = (int) mIter.next();
            int mTmpNumCount = mMap.get(mTmpNum);
            if (mMap.containsKey(mTmpNum + 1)) {
                mResult = Math.max(mResult, mTmpNumCount + mMap.get(mTmpNum + 1));
            }

        }
        return mResult;
    }
}

File: Golang_src/easy/longestHarmoniousSubsequence_594.go
package easy

// Use hashmap:
// Time Complexity: O(n)
// Space Complexity:O(n)
// Runtime: 56 ms, faster than 82.14%
// Memory Usage: 6.7 MB, less than 71.43%
func findLHS(nums []int) int {
	result := 0
	hashmap := make(map[int]int)

	for _, v := range nums {
		hashmap[v]++
	}

	for k, v := range hashmap {
		if hashmap[k+1] > 0 {
			result = max(result, v+hashmap[k+1])
		}
	}

	return result
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// Use hashmap for one loop:
// Time Complexity: O(n)
// Space Complexity:O(n)
// Runtime: 64 ms, faster than 39.29%
// Memory Usage: 6.7 MB, less than 71.43%
func findLHS2(nums []int) int {
	result := 0
	hashmap := make(map[int]int)

	for _, v := range nums {
		hashmap[v]++
		if hashmap[v+1] > 0 {
			result = max(result, hashmap[v]+hashmap[v+1])
		}
		if hashmap[v-1] > 0 {
			result = max(result, hashmap[v]+hashmap[v-1])
		}
	}

	return result
}
